src/server/pps/server/kube_event_tail.go

```diff
@@ -0,0 +1,170 @@
+	lw := cache.NewListWatchFromClient(coreV1.RESTClient(), "events", namespace, fields.Everything())
+	r := cache.NewReflector(lw, &v1.Event{}, &s{ctx: ctx}, 0)
+	r.Run(ctx.Done())
+	// 	return nil
+	// }, backoff.NewInfiniteBackOff(), func(err error, d time.Duration) error {
+	// 	log.Error(ctx, "error in pachw run; will retry", zap.Error(err), zap.Duration("retryAfter", d))
+	// 	return nil
+	// })
+}
+func (s *s) logEvent(e *v1.Event) {
+	if e == nil {
+		return
+	}
+	msg := "event without message"
+	if e.Message != "" {
+		msg = e.Message
+	}
+
+	objRef := func(o *v1.ObjectReference) zapcore.ObjectMarshaler {
+		return zapcore.ObjectMarshalerFunc(func(enc zapcore.ObjectEncoder) error {
+			if o.Namespace != e.Namespace {
+				enc.AddString("namespace", o.Namespace)
+			}
+			enc.AddString("name", o.Kind+"/"+o.Name)
+			if p := o.FieldPath; p != "" {
+				enc.AddString("fieldPath", p)
+			}
+			return nil
+		})
+	}
+	log.Info(s.ctx, msg,
+		zap.Object("kube-event", zapcore.ObjectMarshalerFunc(func(enc zapcore.ObjectEncoder) error {
+			var errs error
+			enc.AddString("namespace", e.ObjectMeta.Namespace)
+			enc.AddString("name", e.ObjectMeta.Name)
+			if err := enc.AddObject("involvedObject", objRef(&e.InvolvedObject)); err != nil {
+				errors.JoinInto(&errs, fmt.Errorf("marshal involvedObject: %w", err))
+			}
+			if e.Reason != "" {
+				enc.AddString("reason", e.Reason)
+			}
+			src := e.Source
+			if src.Component != "" {
+				enc.AddString("source.component", src.Component)
+			}
+			if src.Host != "" {
+				enc.AddString("source.host", src.Host)
+			}
+			if t := e.EventTime.Time; !t.IsZero() {
+				enc.AddTime("eventTime", t)
+			}
+			if t := e.FirstTimestamp.Time; !t.IsZero() {
+				enc.AddTime("firstTimestamp", t)
+			}
+			if t := e.LastTimestamp.Time; !t.IsZero() {
+				enc.AddTime("lastTimestamp", t)
+			}
+			if f, l := e.FirstTimestamp, e.LastTimestamp; !f.IsZero() && !l.IsZero() && l.Sub(f.Time) > time.Second {
+				enc.AddDuration("ongoingDuration", l.Sub(f.Time))
+			}
+			if e.Count > 0 {
+				enc.AddInt32("count", e.Count)
+			}
+			if s := e.Series; s != nil {
+				enc.AddInt32("series.count", s.Count)
+				enc.AddTime("series.lastObservedTime", s.LastObservedTime.Time)
+			}
+			if a := e.Action; a != "" {
+				enc.AddString("action", a)
+			}
+			if e.Related != nil {
+				if err := enc.AddObject("related", objRef(e.Related)); err != nil {
+					errors.JoinInto(&errs, fmt.Errorf("marshal related object: %w", err))
+				}
+			}
+			if rc := e.ReportingController; rc != "" {
+				enc.AddString("reportingController", rc)
+			}
+			if ri := e.ReportingInstance; ri != "" {
+				enc.AddString("reportingInstance", ri)
+			}
+			if t := e.Type; t != "" && t != "Normal" {
+				enc.AddString("type", t)
+			}
+			if errs != nil {
+				enc.AddString("marshalErrors", errs.Error())
+			}
+			return nil
+		})),
+	)
+}
+
+// s is a cache.Store that logs events as they are added.
+type s struct {
+	ctx context.Context
```

Comment:
Probably worth adding a comment that it has to close over the context in order to implement `cache.Store`, which does not take one, and something about not letting instances outlive their calling contexts.